[lldb-dap] Change the launch sequence

JDevlieghere opened the pull request:
This PR changes how we treat the launch sequence in lldb-dap.

 - Send the initialized event after we finish handling the initialize
   request, rather than after we finish attaching or launching.
 - Delay handling the launch and attach request until we have handled
   the configurationDone request. The latter is now largely a NO-OP and
   only exists to signal lldb-dap that it can handle the launch and
   attach requests.
 - Delay handling the initial threads requests until we have handled
   the launch or attach request.
 - Make all attaching and launching asynchronous, including when we have
   attach or launch commands. That removes the need to synchronize
   between the request and event thread.

Background: https://discourse.llvm.org/t/reliability-of-the-lldb-dap-tests/86125

ashgti commented on lldb/tools/lldb-dap/Handler/InitializeRequestHandler.cpp:
```diff
@@ -161,20 +161,22 @@ static void EventThreadFunction(DAP &dap) {
+              std::unique_lock<std::mutex> lock(dap.ignore_stop_mutex);
+              if (dap.ignore_next_stop) {
+                DAP_LOG(dap.log, "Ignoring process stop event");
                 SendThreadStoppedEvent(dap);
```
The line above this sounds like we should skip this?

ashgti commented on lldb/tools/lldb-dap/Handler/InitializeRequestHandler.cpp:
```diff
@@ -161,20 +161,22 @@ static void EventThreadFunction(DAP &dap) {
           case lldb::eStateSuspended:
             break;
           case lldb::eStateStopped:
```
May be unrelated but in [`lldb_private::StateIsRunningState`](https://github.com/llvm/llvm-project/blob/4b1f1f77b0bd7085ecb8c9a431a3161666ec0f54/lldb/source/Utility/State.cpp#L68) and [`lldb_private::StateIsStoppedState`](https://github.com/llvm/llvm-project/blob/4b1f1f77b0bd7085ecb8c9a431a3161666ec0f54/lldb/source/Utility/State.cpp#L89C6-L89C39) it looks like lldb considers a few other states as 'Running' or 'Stopped', should we match those here?

ashgti commented on lldb/tools/lldb-dap/DAP.h:
```diff
@@ -417,8 +424,11 @@ struct DAP {
+  /// Queue for all incoming messages.
   std::deque<protocol::Message> m_queue;
+  /// Dedicated queue for launching and attaching.
+  std::deque<protocol::Message> m_launch_attach_queue;
```
I think there is only ever one launch or attach, should this be an optional message instead? Also named `m_pending_launch_or_attach`?

JDevlieghere replied:
I changed it to an optional, but then backed it out again because we also have to delay the initial `threads` request until we've completed the launch or attach. Let me know if you prefer two optionals instead (i.e. `m_pending_launch_or_attach` and `m_pending_threads`).

ashgti replied:
Looking at things now, I think a separate queue makes more sense.

ashgti commented on lldb/tools/lldb-dap/Handler/InitializeRequestHandler.cpp:
```diff
@@ -161,20 +161,22 @@ static void EventThreadFunction(DAP &dap) {
+            {
+              std::unique_lock<std::mutex> lock(dap.ignore_stop_mutex);
+              if (dap.ignore_next_stop) {
+                DAP_LOG(dap.log, "Ignoring process stop event");
```
Just curious, is there always precisely 1 stop event we should ignore? Or could multiple happen? For example, if we're running `attachCommands` or `launchCommands` could more than 1 stop event be triggered?

JDevlieghere replied:
I think that's a good point, and using launch commands, that't entirely out of our control. You could write a sequence of commands that stop the process an arbitrary number of time. This realization is what made me go with the current approach of doing all the launching and attaching in synchronous mode, because we'll never be able to account for this.

JDevlieghere commented:
@kusmour Can you please confirm whether or not there's a reason to run the launch or attach commands in asynchronous mode? It looks like the divergence was introduced when the feature was added (https://reviews.llvm.org/D154028) and I'm not sure it's necessary. My local testing (using the `gdb-remote 3000` example from the README) seems to work as expected in synchronous mode and even seems like it would desirable?

ashgti commented on lldb/tools/lldb-dap/DAP.cpp:
```diff
@@ -893,10 +893,23 @@ llvm::Error DAP::Loop() {
-            disconnecting = true;
+                  std::get_if<protocol::Request>(&*next)) {
+            if (req->command == "disconnect")
+              disconnecting = true;
```
This will set `disconnecting` to `true` before we've actually responded to the request and this also changed the loop to break as soon as it sees this set to `true` (line 964). If there are multiple queued requests then we'll exit before responding to those requests.

Should we instead wait until the queue is empty instead of `disconnecting` is true on line 964?

JDevlieghere replied:
Ack, I noticed the same thing when running the tests and all of them would fail because we didn't respond to the disconnecting request. I think what you're suggesting makes sense: if we stop reading new messages after we've seen disconnecting, we can use the queue being empty as a way to signal that we're done.

ashgti commented on lldb/tools/lldb-dap/DAP.cpp:
```diff
@@ -893,10 +893,23 @@ llvm::Error DAP::Loop() {
+          // be answered until we've gotten the confgigurationDone request. We
+          // can't answer the threads request until we've successfully launched
+          // or attached.
+          bool is_part_of_launch_sequence = false;
```
After `request initialize` + `event initialized`, should we queue everything until we see `configurationDone`? We're already making a special case for `threads` and I worry about other special cases. Maybe the general flow should be to put all messages into the launch queue except `initialize`, `configurationDone` and `disconnect` (in case we need to shutdown before launching is complete) until we've responded to `configurationDone`. Then we can process the launch queue and processed as normal at that point.

ashgti replied:
We don't have to respond to those requests immediately, they're sent, but at least VSCode doesn't wait for a response to them before sending the `configurationDone` request.

ashgti replied:
I guess that would be a problem though because we'd miss hitting initial breakpoints, so ya, your right about those requests.

ashgti commented on lldb/tools/lldb-dap/DAP.cpp:
```diff
@@ -938,15 +954,30 @@ llvm::Error DAP::Loop() {
+    m_queue_cv.wait(lock, [&] {
+      return disconnecting || !m_queue.empty() ||
+             (!m_launch_sequence_queue.empty() && configuration_done);
+    });
 
-    if (m_queue.empty())
+    if (disconnecting)
       break;
 
-    Message next = m_queue.front();
-    m_queue.pop_front();
+    bool is_launch_seqeuence =
+        !m_launch_sequence_queue.empty() && configuration_done;
+
+    auto &active_queue =
+        is_launch_seqeuence ? m_launch_sequence_queue : m_queue;
+
+    assert(!active_queue.empty() &&
+           "shouldn't have gotten past the wait with an empty queue");
```
What if we kept the messages in the `m_launch_sequence_queue` until `configurationDone` and in `configurationDone` `PostRun()` we move the items in the `m_launch_sequence_queue` into the front of `m_queue`?

Then this loop would only need to check the `m_queue.empty()` and have a few less branching paths.

We could add a helper like:

```
void DAP::SetConfigurationDone() {
  std::lock_guard<std::mutex> guard(m_queue_mutex);
  std::copy(m_launch_sequence_queue.front(), m_launch_sequence_queue.end(), std::front_inserter(m_queue));
  configuration_done = true;
}
```

and call in `ConfigurationDoneRequestHandler`.

ashgti commented on lldb/packages/Python/lldbsuite/test/tools/lldb-dap/dap_server.py:
```diff
@@ -591,6 +591,7 @@ def request_attach(
         attachCommands=None,
         terminateCommands=None,
         coreFile=None,
+        stopOnAttach=True,
```
Should we default to `False` for attach requests?

JDevlieghere commented:
> > @kusmour Can you please confirm whether or not there's a reason to run the launch or attach commands in asynchronous mode? It looks like the divergence was introduced when the feature was added (https://reviews.llvm.org/D154028) and I'm not sure it's necessary. My local testing (using the `gdb-remote 3000` example from the README) seems to work as expected in synchronous mode and even seems like it would desirable?
> 
> I just synced with Greg on some context. I think the nature of attachCommands will allow people do anything. What if they have a waitFor command that will not get timed out by GDBRemote timeout? Then lldb-dap will never return anything to the client. The `dap.WaitForProcessToStop(timeout_seconds);` will guarantee we at least return error if things aren't working.

Right, that's exactly the problem, and it goes both ways. Similarly to the situation where there's no stop, we have the same problem if there's more than one stop. 

> In fact maybe we should consider making the normal launch/attach in async and have the `dap.WaitForProcessToStop(timeout_seconds);` at the end to ensure the process is in expected state before responding to the launch/attach req?

That's what I wanted to do originally, but I think that makes things strictly more complicated because it requires lldb-dap to deal with all the stops. Imagine you have a launch command that stops and resumes the inferior more than once. If we do that asynchronously, we'll get multiple stop events. 

 - If stop-on-entry is false, we have to issue the continue so the process is running when the request completes. How do you know after which stop event to issue the continue? Because the process events are delivered asynchronously, you don't know when you're done handling all the events belonging to the launch or attach. 
 - If stop-on-entry is false, we have to make sure the process is in a stopped state when the request completes. How do you know after which running event to issue the continue? 

At this point I'm convinced that there's no sound way of supporting all these scenarios from lldb-dap. The benefit of using synchronous mode is that we can let lldb itself deal with most of the stops, at the cost of potentially not doing the right thing when you have launch commands. I hope I'm wrong or missing something, but I would really like to solve this better, but I'm out of ideas.

ashgti commented on lldb/packages/Python/lldbsuite/test/tools/lldb-dap/lldbdap_testcase.py:
```diff
@@ -333,6 +333,7 @@ def attach(
         exitCommands=None,
         attachCommands=None,
         coreFile=None,
+        stopOnAttach=True,
```
Should we default this to `False`?

Or should we have the `DAPTestCaseBase` case `attach` and `launch` helpers take an `initialSourceBreakpoints, initialFunctionBreakpoints, initialExceptionBreakpoints` parameter and we help the tests by sending these as part of the setup flow, kind of how we have `disconnectAutomatically` as a flag to handle shutdown flows automatically.

JDevlieghere replied:
The way the tests are currently written, they expect to be stopped after the launch (presumably because these tests were never sending the `configurationDone` request). I went with this default to keep the changes minimal, but I like the idea of the initial breakpoints as a way to improve this in a subsequent PR.

ashgti replied:
Sounds good, we can follow up with some other improvements to tests.

kusmour commented on lldb/packages/Python/lldbsuite/test/tools/lldb-dap/dap_server.py:
```diff
@@ -591,6 +591,7 @@ def request_attach(
         attachCommands=None,
         terminateCommands=None,
         coreFile=None,
+        stopOnAttach=True,
```
Is there any reason we want to make this true by default?
1) Our current tests don't have stopOnEntry by default
2) DAP doesn't have definition about this nor does lldb-dap documentation. So most real life attach case will not have this turned on.
3) I don't think we have lldb-dap test case covering multithreading scenario. But on linux, ptrace will put SIGSTOP on every thread and currently in `SendThreadStoppedEvent(dap)` we will send a `stopped` event for any threads that has the a stop reason, meaning we will actually send N stopped event (N being # of threads) upon attach request with `stopOnAttach` set to true.

Though it's only turned on for tests, I think we should make it false until it become the expected default behavior.

JDevlieghere replied:
See my reply to John: https://github.com/llvm/llvm-project/pull/138219#discussion_r2072197289

kusmour commented on lldb/packages/Python/lldbsuite/test/tools/lldb-dap/lldbdap_testcase.py:
```diff
@@ -357,6 +358,7 @@ def cleanup():
         self.addTearDownHook(cleanup)
         # Initialize and launch the program
         self.dap_server.request_initialize(sourceInitFile)
+        self.dap_server.request_configurationDone()
```
Shouldn't we wait for `initialized` event?
```suggestion
        self.dap_server.wait_for_event("initialized")
        self.dap_server.request_configurationDone()
```

kusmour commented on lldb/packages/Python/lldbsuite/test/tools/lldb-dap/lldbdap_testcase.py:
```diff
@@ -427,6 +430,7 @@ def cleanup():
 
         # Initialize and launch the program
         self.dap_server.request_initialize(sourceInitFile)
+        self.dap_server.request_configurationDone()
```
ditto

kusmour commented on lldb/test/API/tools/lldb-dap/attach/TestDAP_attach.py:
```diff
@@ -25,9 +25,10 @@ def spawn_and_wait(program, delay):
-@skipIf
 class TestDAP_attach(lldbdap_testcase.DAPTestCaseBase):
     def set_and_hit_breakpoint(self, continueToExit=True):
+        self.dap_server.wait_for_stopped()
```
Please see the comment about default `stopOnAttach`

kusmour commented on lldb/test/API/tools/lldb-dap/completions/TestDAP_completions.py:
```diff
@@ -235,7 +235,7 @@ def test_auto_completions(self):
         Tests completion requests in "repl-mode=auto"
         """
-        self.setup_debugee()
+        self.setup_debugee(stopOnEntry=True)
```
Any reason why we are changing the test behavior. I believe currently they will be false.

kusmour commented on lldb/test/API/tools/lldb-dap/launch/TestDAP_launch.py:
```diff
@@ -88,8 +88,8 @@ def test_stopOnEntry(self):
         self.build_and_launch(program, stopOnEntry=True)
-        self.set_function_breakpoints(["main"])
-        stopped_events = self.continue_to_next_stop()
+
```
nit: probs should update the line 98 & 99 below
```
                    self.assertEqual(
                        reason, "singal", 'verify stop reason is SIGSTOP'
                    )
```

kusmour commented on lldb/tools/lldb-dap/Handler/AttachRequestHandler.cpp:
```diff
@@ -137,55 +137,69 @@ void AttachRequestHandler::operator()(const llvm::json::Object &request) const {
+  // Make sure the process is attached and stopped.
+  error = dap.WaitForProcessToStop(std::chrono::seconds(timeout_seconds));
+
+  // Clients can request a baseline of currently existing threads after
+  // we acknowledge the configurationDone request.
+  // Client requests the baseline of currently existing threads after
+  // a successful or attach by sending a 'threads' request
+  // right after receiving the configurationDone response.
+  // Obtain the list of threads before we resume the process
+  dap.initial_thread_list =
+      GetThreads(dap.target.GetProcess(), dap.thread_format);
```
If we delay the handling of launch/attach request after responding to `configurationDone`, what happens to `threads` request? Is it possible that lldb-dap gets a `threads` request before saving the initial_thread_list here?

VSCode [sends out threads request right after the `configurationDone` response](https://github.com/microsoft/vscode/blob/174af221c9ea2ccdb64abe4aab8e1a805e77beae/src/vs/workbench/contrib/debug/browser/debugSession.ts#L1088-L1089)

JDevlieghere replied:
The threads request gets delayed as well. As long as it comes in after the launch/attach request, it will get handled after. Here's the log between VS Code and the implementation in this PR:

```
--> {"command":"initialize","arguments":{"clientID":"vscode","clientName":"Visual Studio Code","adapterID":"lldb-dap","pathFormat":"path","linesStartAt1":true,"columnsStartAt1":true,"supportsVariableType":true,"supportsVariablePaging":true,"supportsRunInTerminalRequest":true,"locale":"en","supportsProgressReporting":true,"supportsInvalidatedEvent":true,"supportsMemoryReferences":true,"supportsArgsCanBeInterpretedByShell":true,"supportsMemoryEvent":true,"supportsStartDebuggingRequest":true,"supportsANSIStyling":true},"type":"request","seq":1}
<-- {"body":{"$__lldb_version":"lldb version 21.0.0git (git@github.com:JDevlieghere/llvm-project.git revision e7333489e50b1efe16203e8fd7b6dc7f2dcd4439)\n  clang revision e7333489e50b1efe16203e8fd7b6dc7f2dcd4439\n  llvm revision e7333489e50b1efe16203e8fd7b6dc7f2dcd4439","completionTriggerCharacters":["."," ","\t"],"exceptionBreakpointFilters":[{"default":false,"filter":"cpp_catch","label":"C++ Catch"},{"default":false,"filter":"cpp_throw","label":"C++ Throw"},{"default":false,"filter":"objc_catch","label":"Objective-C Catch"},{"default":false,"filter":"objc_throw","label":"Objective-C Throw"}],"supportTerminateDebuggee":true,"supportsBreakpointLocationsRequest":true,"supportsCancelRequest":true,"supportsCompletionsRequest":true,"supportsConditionalBreakpoints":true,"supportsConfigurationDoneRequest":true,"supportsDataBreakpoints":true,"supportsDelayedStackTraceLoading":true,"supportsDisassembleRequest":true,"supportsEvaluateForHovers":true,"supportsExceptionInfoRequest":true,"supportsExceptionOptions":true,"supportsFunctionBreakpoints":true,"supportsHitConditionalBreakpoints":true,"supportsInstructionBreakpoints":true,"supportsLogPoints":true,"supportsModulesRequest":true,"supportsReadMemoryRequest":true,"supportsRestartRequest":true,"supportsSetVariable":true,"supportsStepInTargetsRequest":true,"supportsSteppingGranularity":true,"supportsValueFormattingOptions":true},"command":"initialize","request_seq":1,"seq":0,"success":true,"type":"response"}
<-- {"event":"initialized","seq":0,"type":"event"}
--> {"command":"launch","arguments":{"type":"lldb-dap","request":"launch","name":"Launch LLDB","program":"/Users/jonas/llvm/build-ra/bin/lldb","args":[],"env":[],"cwd":"/Users/jonas/llvm/llvm-project","__configurationTarget":6,"__sessionId":"8dc997cc-ed4f-4ce1-a7e8-f2d59ea84796"},"type":"request","seq":2}
--> {"command":"setFunctionBreakpoints","arguments":{"breakpoints":[]},"type":"request","seq":3}
<-- {"body":{"breakpoints":[]},"command":"setFunctionBreakpoints","request_seq":3,"seq":0,"success":true,"type":"response"}
--> {"command":"setInstructionBreakpoints","arguments":{"breakpoints":[]},"type":"request","seq":4}
--> {"command":"setExceptionBreakpoints","arguments":{"filters":[]},"type":"request","seq":5}
<-- {"body":{"breakpoints":[]},"command":"setInstructionBreakpoints","request_seq":4,"seq":0,"success":true,"type":"response"}
--> {"command":"setDataBreakpoints","arguments":{"breakpoints":[]},"type":"request","seq":6}
<-- {"command":"setExceptionBreakpoints","request_seq":5,"seq":0,"success":true,"type":"response"}
<-- {"body":{"breakpoints":[]},"command":"setDataBreakpoints","request_seq":6,"seq":0,"success":true,"type":"response"}
--> {"command":"configurationDone","type":"request","seq":7}
<-- {"command":"configurationDone","request_seq":7,"seq":0,"success":true,"type":"response"}
--> {"command":"threads","type":"request","seq":8}
<-- {"body":{"module":{"addressRange":"4294967296","debugInfoSize":"","id":"CFAF6A92-9494-3549-9408-121B5F8E5A32","name":"lldb","path":"/Users/jonas/llvm/build-ra/bin/lldb","symbolFilePath":"/Users/jonas/llvm/build-ra/bin/lldb","symbolStatus":"Symbols loaded."},"reason":"removed"},"event":"module","seq":0,"type":"event"}
<-- {"command":"launch","request_seq":2,"seq":0,"success":true,"type":"response"}
<-- {"body":{"isLocalProcess":true,"name":"/Users/jonas/llvm/build-ra/bin/lldb","startMethod":"launch","systemProcessId":24119},"event":"process","seq":0,"type":"event"}
<-- {"body":{"threads":[{"id":3321319,"name":"Thread 1"}]},"command":"threads","request_seq":8,"seq":0,"success":true,"type":"response"}
```

ashgti reviewed (approved):
Yesterday I enabled all the lldb-dap tests that support macOS and I noticed #138197 where some tests `attach` failed due to attaching to a parallel test process. I was consistently hitting it on my MBP but I think due to timing not on my Mac Studio (different number of cores, I think).

JDevlieghere commented:
@kusmour Please take another look. In the current state all the test pass (reliably) locally and the linux bot seems happy too. Not sure if we want to keep the test disabled for now or turn them back on and see what the reliability is like on the bots.

kusmour commented:
Unblocked!

> @kusmour Please take another look. In the current state all the test pass (reliably) locally and the linux bot seems happy too. Not sure if we want to keep the test disabled for now or turn them back on and see what the reliability is like on the bots.

I don't know much about the build bot infra but looks like the only way to get the tests running is to land an enablement patch?

Looks like a bunch of tests got disabled today: https://github.com/llvm/llvm-project/issues/137660
I don't think we should keep ignoring tests if we want to land this tho. I am all for re-enable them with the risk of reverting. But I will let others to chime in.

cc. @IanWood1 @felipepiovezan

JDevlieghere commented:
> Unblocked!

Thanks!

> > @kusmour Please take another look. In the current state all the test pass (reliably) locally and the linux bot seems happy too. Not sure if we want to keep the test disabled for now or turn them back on and see what the reliability is like on the bots.
> 
> I don't know much about the build bot infra but looks like the only way to get the tests running is to land an enablement patch?

We  have pre-commit  testing (which passed) but that doesn't say much about the reliability. We'll need to enable them to see what the impact is on the post-commit CI. I think I'll split off the test-reenablement into a separate PR so we don't have to revert the whole thing here.

> Looks like a bunch of tests got disabled today: #137660 I don't think we should keep ignoring tests if we want to land this tho. I am all for re-enable them with the risk of reverting. But I will let others to chime in.
> 
> cc. @IanWood1 @felipepiovezan

I don't t think those got disabled today, they're cherrypicks on forks from the upstream commits (the ones I mentioned on Discouse):

- [[lldb][test] Disable a bunch of flakey lldb-dap tests · llvm/llvm-project@2dbab4c · GitHub](https://github.com/llvm/llvm-project/commit/2dbab4ca8ddb218af555d8d1fd86b72612387582)
- [[lldb] Disable TestDAP_attachByPortNum · llvm/llvm-project@b73169a · GitHub](https://github.com/llvm/llvm-project/commit/b73169ad3edc56fe2b4f5bd3d5259310168da50c)
- [[lldb] Disable TestDAP_attach · llvm/llvm-project@55b4e5e · GitHub](https://github.com/llvm/llvm-project/commit/55b4e5e5154550ba981af08ca9bd1e3da00e6fea)

JDevlieghere committed:
[lldb-dap] Change the launch sequence

This PR changes how we treat the launch sequence in lldb-dap.

 - Send the initialized event after we finish handling the initialize
   request, rather than after we finish attaching or launching.
 - Delay handling the launch and attach request until we have handled
   the configurationDone request. The latter is now largely a NO-OP and
   only exists to signal lldb-dap that it can handle the launch and
   attach requests.
 - Delay handling the initial threads requests until we have handled
   the launch or attach request.
 - Make all attaching and launching asynchronous, including when we have
   attach or launch commands. That removes the need to synchronize
   between the request and event thread.

Background: https://discourse.llvm.org/t/reliability-of-the-lldb-dap-tests/86125

JDevlieghere commented:
I created #138778 to track extending the launch and attach helpers to take breakpoints.

DavidSpickett commented:
We have failing tests on Windows after this. Latest result at this time is:
```
********************
Unresolved Tests (2):
  lldb-api :: tools/lldb-dap/completions/TestDAP_completions.py
  lldb-api :: tools/lldb-dap/startDebugging/TestDAP_startDebugging.py
********************
Timed Out Tests (1):
  lldb-api :: tools/lldb-dap/send-event/TestDAP_sendEvent.py
********************
Failed Tests (6):
  lldb-api :: tools/lldb-dap/console/TestDAP_console.py
  lldb-api :: tools/lldb-dap/console/TestDAP_redirection_to_console.py
  lldb-api :: tools/lldb-dap/launch/TestDAP_launch.py
  lldb-api :: tools/lldb-dap/stackTrace/TestDAP_stackTrace.py
  lldb-api :: tools/lldb-dap/stackTraceDisassemblyDisplay/TestDAP_stackTraceDisassemblyDisplay.py
  lldb-api :: tools/lldb-dap/variables/children/TestDAP_variables_children.py
```
Which includes some follow ups.

DavidSpickett commented:
I've reverted this. Please take a look at the logs of https://lab.llvm.org/buildbot/#/builders/141/builds/8500 and see if any of it makes sense to you.

We didn't have failures on Linux (though a lot of tests are disabled, they would be disabled everywhere) so my first instinct would be any construct that might be different on Windows.

JDevlieghere commented:
The following two tests are both setting breakpoints without stopping at entry, which means that we could have run past the breakpoint by the time it's set:

```
  lldb-api :: tools/lldb-dap/completions/TestDAP_completions.py
  lldb-api :: tools/lldb-dap/startDebugging/TestDAP_startDebugging.py
```

labath commented:
I think I've found a(nother) problem with this change. One of the arguments of the launch request is `sourceMap`, which sets the `target.source-map` setting. The setting is necessary to correctly resolve file+line breakpoints in case the file's path on the host system does not match the path from the debug info. By deferring the processing of this request, we're effectively ignoring this setting as all of the initial breakpoint requests (which are not deferred) will not see the intended value. This is kind of a Big Deal(tm) for us since all of our binaries are built in the cloud.

I see this has been kind of reverted in #140331. I suspect that fixes this use case (checking that is my next step), but I'm writing this to let you know of the problem.

labath commented:
That would be great, thanks.

BTW, I've checked the test and its form after https://github.com/llvm/llvm-project/pull/140331 (where `self.launch` does not send the configurationDone` event) does reproduce the problem I encountered.
